File: 01_GOF23/DP_examples/Go/02_Structural/Proxy/proxy.go
package Proxy

// ImageProxy 是图像代理类，实现了 Image 接口
type ImageProxy struct {
	Images []Image
}

// NewImageProxy 创建一个新的 ImageProxy 实例
func NewImageProxy() *ImageProxy {
	return &ImageProxy{
		Images: make([]Image, 0),
	}
}

// Operation 执行所有图像的操作
func (p *ImageProxy) Operation() {
	for _, image := range p.Images {
		image.Operation()
	}
}

// LoadFromDisk 加载所有图像
func (p *ImageProxy) LoadFromDisk() {
	for _, image := range p.Images {
		image.LoadFromDisk()
	}
}

// AddImage 添加单个图像
func (p *ImageProxy) AddImage(image Image) {
	p.Images = append(p.Images, image)
}

// AddImages 添加多个图像
func (p *ImageProxy) AddImages(images ...Image) {
	p.Images = append(p.Images, images...)
}

// Remove 移除图像
func (p *ImageProxy) Remove(image Image) bool {
	for i, img := range p.Images {
		if img == image {
			p.Images = append(p.Images[:i], p.Images[i+1:]...)
			return true
		}
	}
	return false
}

File: 01_GOF23/DP_examples/Lua/02_Structural/Proxy/Proxy.lua
-- 加载 Image 接口
require("RealSubject")

-- ImageProxy 是图像代理类，实现了 Image 接口
---@class ImageProxy : Image
---@field images table
ImageProxy = Image:new()
function ImageProxy:new()
    local o = Image.new(self)
    o.images = {}
    return o
end
function ImageProxy:operation()
    for _, image in ipairs(self.images) do
        image:operation()
    end
end

-- 加载所有图像
function ImageProxy:loadFromDisk()
    for _, image in ipairs(self.images) do
        image:loadFromDisk()
    end
end

-- 添加单个图像
function ImageProxy:addImage(image)
    table.insert(self.images, image)
end

-- 添加多个图像
function ImageProxy:addImages(...)  
    for _, image in ipairs({...}) do
        table.insert(self.images, image)
    end
end

-- 移除图像
function ImageProxy:remove(image)
    for i, img in ipairs(self.images) do
        if img == image then
            table.remove(self.images, i)
            return true
        end
    end
    return false
end
